### PR TITLE
DateTime types implicit conversion from varchar

### PR DESCRIPTION
**Pull request by fiedukow**
Additionaly TIME_WITH_TIME_ZONE from VARCHAR missing conversion operator was implemented.

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/type/TestDate.java [hunk 1]
```diff
@@ -38,12 +39,11 @@
 import static org.joda.time.DateTimeZone.UTC;
 
 public class TestDate
+    extends AbstractTestFunctions
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/type/TestDate.java]: Indentation is wrong for all of these. Should be 8 spaces

File: presto-main/src/test/java/com/facebook/presto/type/TestDate.java [hunk 2]
```diff
@@ -193,4 +188,42 @@ public void testLeast()
...
+        assertEvaluates("'2000-01-01' <> DATE '2000-01-01'", BOOLEAN);
+        assertEvaluates("'2000-01-01' <= DATE '2000-01-01'", BOOLEAN);
+        assertEvaluates("'2000-01-01' >= DATE '2000-01-01'", BOOLEAN);
+    }
+
+    @Test
+    public void dateIsImplicitlyCastedForMostSpecificFunction()
+    {
+        // Test if there is a possibility of solving ambiguity by selecting
+        // most specific function it's done for date format varchargit .
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/type/TestDate.java]: Typo "varchargit"

**Comment by electrum** [comment]
Please split out the missing conversion into a separate commit.

**Comment by electrum** [comment]
While convenient, however, I have concerns that this is not safe and can introduce silent errors into queries. Is this required by the SQL specification?

**Comment by electrum** [comment]
We don't implicitly convert varchar to other types like numbers. Why are dates different?

One problem is that, unlike conversions such as bigint to double, a cast from a varchar to a date can fail due to invalid data. This means an expression like `x = y` might work fine, then suddenly start failing if invalid data is introduced. It's not at all obvious looking at the expression that a cast is occurring.

**Comment by martint** [comment]
Yeah, I don't think this is something we want to add. Implicit coercions should be "widening" operations, so they should not fail on input. If anything, the opposite is the one that could be allowed, but whether it's a good idea to even have that is debatable.

**Comment by martint** [comment]
It also results in strange side-effects, like allowing an expression such as '2016-01-01' - interval '1' day

**Comment by fiedukow** [comment]
Starting with short clarification:
 1. As for silent fails - I think there are no silent-fail scenarios in here. If something goes terribly wrong it will cause an exception. Silent errors are possible due to coersions between multiple DateTime types, but it is not something introduced here.
 2. "We don't implicitly convert varchar to other types like numbers. Why are dates different?" - the most important difference is that all numeric types do have a way to be implicitly constructed, date requires explicit constructor.
 3. In principle the query:
    `'2016-01-01' - interval '1' day`
    is ok and will coerce after this changes. However as there is ambiguity between TIME and DATE in first argument, this query will NOT pass after this changes, instead it will suggest two candidates:
`presto> select '2016-01-01' - interval '1' day;
Query 20160601_084740_00020_8fkwv failed: line 1:21: Could not choose a best candidate operator. Explicit type casts must be added.
Candidates are:
	 * $operator$SUBTRACT(date,interval day to second):date
	 * $operator$SUBTRACT(time,interval day to second):time
`

Now as for pro and cons for this change i see at this point:
Cons:
 - **Queries using those coersions may fail or not depending on input data.** - the only other example (that i know about) of such behavior in presto are Regexp types, but those types are 'special'.
 - Once introduced the implicit coersion is hard to get rid of.
Pro:
 + It's widely used practice and for that reason it allows to run a lot of SQL queries prepared for other DBs without changes.
 + SQL is less verbose making using REPL much simpler.

As for the timestamp -> varchar(x) coercion, I think it's actually much worse - losing well organized data in favor of general purpose VARCHAR(x) seems like something that user should always do consciously.

So, leaving the discussion to people that are probably more convinced that this is a good move ;-)
Any thoughts on the topic @arhimondr @KBP-TDC ?

**Comment by fiedukow** [comment]
Requested changes introduced.
Commit split performed.

**Comment by fiedukow** [comment]
Bump!
How about we tackle the problem from other side and just introduce comparison operators between VARCHAR and DATE/TIME types?
This will allow us to avoid problems with coersion in unwanted contexts and allow to support the comparison DATE to VARCHAR which is how it works in at least Oracle DB, PostgreSQL and MySQL.
Such comparisons are also performed in TPC-H benchmark so it would be nice to support this - as it looks like a standard by example.

**Comment by electrum** [comment]
What if we added implicit constructors for constants (but not generic coercions)? Would that capture much of the benefit without introducing problems?

**Comment by martint** [comment]
Sorry, I haven't had a chance to look at this further, yet. Which tpch queries need this?

**Comment by martint** [comment]
@electrum, that seems like a bad idea. I makes the type system inconsistent and confusing. For instance, while this might be valid:

```sql
SELECT foo('a')
```

this would not be:

```sql
SELECT foo(x) FROM (SELECT 'a' x)

SELECT foo(x) FROM (VALUES 'a') t(x)
```

Another example. This would be valid:

```sql
SELECT x = '1' FROM (VALUES 1) t(x)
```

But this would be ambiguous:

```sql
SELECT 1 = '1'
```

It would also allow expressions like:

```
true AND 'true' AND 1 AND 1.0
foo['1']
123 LIKE 456
```

**Comment by KBP-TDC** [comment]
@martint Regarding TPC-H, I just verified that the latest version of the spec uses `date '1998-12-01'` convention so we are good there. However, users often go for shortcuts offered by most of other DBMS/SQL engines: 
```
SELECT COUNT(*)
FROM lineitem
WHERE l_shipdate <= '1998-12-01'
``` 
as @fiedukow mentioned earlier. 
Is there an acceptable way to implement that in Presto?

**Comment by fiedukow** [comment]
ping, any new thoughts on the topic?

**Comment by fiedukow** [comment]
rebased to master, resolved conflicts

**Comment by fiedukow** [comment]
rebased to current master, resolved conflicts.
Still waiting for clue about this PR future and your position on our proposition of solving the use case @electrum @martint

**Comment by fiedukow** [comment]
@martint @cberner The solution I proposed on Jun 9 is of course not acceptable after we decided that we are burring idea of having comparison between different types for now.
The question now is, if you see any chance or possibility of this syntax landing on master?

Just to recap the main focus here is to make literals like '2000-01-01' to be used as the DATE/TIME/etc. More details are in previous comments.

**Comment by cberner** [comment]
I don't think this is a good idea. There are a number of problems such as `'1-1-1' != '1-1-01'` but `date '1-1-1' = date '1-1-01'`. Also, I don't think there's a good argument for why the coercion goes from varchar to date/time/... rather than the other way around, which means that there's no intuitive answer to `'1-1-1' = date '1-1-1'` (it's false in one case, but true in the other)

**Comment by fiedukow** [comment]
I don't thing changes are requested here. 
I would rather close this as rejected. If you don't think it's right solution please reopen.

**Pull request closed without merging**